Epsilon/Service/FilterService.cs

```diff
@@ -0,0 +1,62 @@
+using DocumentFormat.OpenXml;
+using DocumentFormat.OpenXml.Packaging;
+using DocumentFormat.OpenXml.Wordprocessing;
+using Epsilon.Abstractions.Component;
+using Epsilon.Abstractions.Service;
+using Epsilon.Canvas.Abstractions.Model;
+using Epsilon.Canvas.Abstractions.Model.GraphQl;
+using Epsilon.Canvas.Abstractions.Service;
+using Microsoft.Extensions.Configuration;
+
+namespace Epsilon.Service;
+
+public class FilterService : IFilterService
+{
+    private const string GetUserSubmissionsQuery = @"
+        query GetUserSubmissions {
+          allCourses {
+            submissionsConnection(studentIds: $studentIds) {
+              nodes {
+                submittedAt
+              }
+            }
+          }
+        }
+    ";
+
+    private readonly IConfiguration _configuration;
+    private readonly IAccountHttpService _accountHttpService;
+    private readonly IGraphQlHttpService _graphQlService;
+
+    public FilterService(IConfiguration configuration, IAccountHttpService accountHttpService, IGraphQlHttpService graphQlHttpService)
+    {
+        _configuration = configuration;
+        _accountHttpService = accountHttpService;
+        _graphQlService = graphQlHttpService;
+    }
+
+    public async Task<IEnumerable<EnrollmentTerm>> GetParticipatedTerms()
+    {
+        var allTerms = await _accountHttpService.GetAllTerms(1);
+
+        var studentId = _configuration["Canvas:StudentId"];
+        var submissionsQuery = GetUserSubmissionsQuery.Replace("$studentIds", $"{studentId}", StringComparison.InvariantCultureIgnoreCase);
+
+        var response = await _graphQlService.Query<CanvasGraphQlQueryResponse>(submissionsQuery);
+
+        if (response!.Data == null)
```

Comment:
If you use an exclamation mark here then it will still throw a massive exception when response is null. I would go for the ? operator. It accesses the Data property only when response is not null, so does 2 null checks basically. https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-

Reply:
Thanks for the notice, I'll change it up!